email/email.go

```diff
@@ -0,0 +1,311 @@
+		txn, emailAddress, publicKey.Fingerprint(),
+		meta.RequestUserAgent,
+		meta.RequestIpAddress,
+		meta.RequestTime,
+	)
+	if err != nil {
+		return err
+	}
+
+	emailTemplateData := verifyEmail{
+		Email:            emailAddress,
+		VerificationUrl:  makeVerificationUrl(*verifySecretUUID),
+		RequestIpAddress: meta.RequestIpAddress,
+		RequestTime:      meta.RequestTime,
+		KeyFingerprint:   publicKey.Fingerprint().Hex(),
+		KeyCreatedDate:   publicKey.PrimaryKey.CreationTime,
+	}
+
+	email := email{
+		to:      emailAddress,
+		from:    "Fluidkeys <verify@mail.fluidkeys.com>",
+		replyTo: "Fluidkeys Security <security@fluidkeys.com>",
+		bcc:     "hello@fluidkeys.com",
+	}
+
+	if err := email.renderSubjectAndBody("verify", emailTemplateData); err != nil {
+		return fmt.Errorf("error rendering email: %v", err)
+	}
+
+	log.Printf("email.htmlBody: %s\n", email.htmlBody)
+
+	if err := email.send(); err != nil {
+		return fmt.Errorf("error sending mail: %v", err)
+	}
+	return nil
+}
+
+// shouldSendVerificationEmail returns true if an email address should receive a new verification
+// email
+func shouldSendVerificationEmail(txn *sql.Tx, email string) (bool, error) {
+	_, alreadyLinked, err := datastore.GetArmoredPublicKeyForEmail(email)
+	if err != nil {
+		return false, err
+	}
+	if alreadyLinked {
+		// 1. it's linked to the same key, in which case there's
+		//    nothing to do
+		// 2. it's linked to one key, and this is a request
+		//    to link the email to a *different* key, which we don't
+		//    currently allow. The email_key_link must be deleted
+		//    before the email can be linked again. Note that this
+		//    happens if the whole linked *key* is deleted.
+		return false, nil
+	}
+
+	hasActiveVerification, err := datastore.HasActiveVerificationForEmail(txn, email)
+	if err != nil {
+		return false, err
+	}
+	if hasActiveVerification {
+		// prevents an attacker from mailbombing an email address by
+		// creating lots of different keys with the email: if there's
+		// an active email verification, it has to expire before
+		// another one can be created
+		log.Printf("email verification already exists for %s, not sending another", email)
+		return false, nil
+	}
+
+	log.Printf("no currently-active verifications for email '%s'", email)
+
+	return true, nil
+}
+
+func makeVerificationUrl(secretUUID uuid.UUID) string {
+	return fmt.Sprintf("https://api.fluidkeys.com/v1/emails/verify/%s", secretUUID.String())
+}
+
+type email struct {
+	to       string
+	from     string
+	replyTo  string
+	bcc      string
+	subject  string
+	htmlBody string
+}
+
+func inferTemplateName(emailTemplateData interface{}) (string, error) {
+	switch emailTemplateData.(type) {
+	case verifyEmail:
+		return "verify", nil
+	}
+
+	return "", fmt.Errorf("failed to get template name from data: %v", emailTemplateData)
+}
+
+func (e *email) renderSubjectAndBody(templateName string, data interface{}) (err error) {
+	templateName, err = inferTemplateName(data)
```

Comment:
As far as I can see you're passing `verify` into `templateName`, why bother trying to `inferTemplateName` aswell?

Reply:
Good spot!